### PR TITLE
Add io.github.picocrypt.Picocrypt

### PR DESCRIPTION
**Pull request by HACKERALERT**
### Please confirm your submission meets all the criteria

- [X] Please describe the application briefly.
A very small, very simple, yet very secure graphical file encryption tool written in Go.

- [X] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [X] I have read the and followed all the [Submission and licence requirements][reqs].
- [X] I have [built][build] and tested the submission locally.
- [X] I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about this submission. **Link:**

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

### CODE REVIEW

File: flathub.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "only-arches": ["x86_64"]
```
Comment (hfiguiere) [inline flathub.json:2]: why? Go runs on aarch64 too

File: io.github.picocrypt.Picocrypt.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      - install -Dm644 io.github.picocrypt.Picocrypt.metainfo.xml /app/share/metainfo/io.github.picocrypt.Picocrypt.metainfo.xml
+    sources:
+      - type: file
+        url: https://github.com/Picocrypt/Picocrypt/releases/download/1.42/Picocrypt
+        sha256: 7112b72d0b62337f9249e47f6079b8a041c445fc089f0610ca3af040514e4ab2
+      - type: file
+        url: https://raw.githubusercontent.com/Picocrypt/Picocrypt/main/images/key.svg
+        sha256: f2f3b9ad8322d8e0d9bc3ba888fa89310bfe54800f48d53a4260c71c96f79cf6
+      - type: file
+        url: https://raw.githubusercontent.com/Picocrypt/flathub/io.github.picocrypt.Picocrypt/io.github.picocrypt.Picocrypt.metainfo.xml
```
Comment (hfiguiere) [inline io.github.picocrypt.Picocrypt.yaml:36]: no. this reference the wrong place. 

1. it's like so it's not URL
2. it actually should be upstream

File: io.github.picocrypt.Picocrypt.yaml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    buildsystem: simple
+    build-commands:
+      - touch io.github.picocrypt.Picocrypt.desktop
+      - echo "[Desktop Entry]" >> io.github.picocrypt.Picocrypt.desktop
+      - echo "Name=Picocrypt" >> io.github.picocrypt.Picocrypt.desktop
+      - echo "Comment=Secure file encryption tool" >> io.github.picocrypt.Picocrypt.desktop
+      - echo "Type=Application" >> io.github.picocrypt.Picocrypt.desktop
+      - echo "Exec=picocrypt-gui" >> io.github.picocrypt.Picocrypt.desktop
+      - echo "Icon=io.github.picocrypt.Picocrypt" >> io.github.picocrypt.Picocrypt.desktop
+      - echo "Categories=Utility;Security;" >> io.github.picocrypt.Picocrypt.desktop
```
Comment (hfiguiere) [inline io.github.picocrypt.Picocrypt.yaml:23]: (lines 16:23) No. You should have an actual .desktop file. And t shoud also be submitted upstream. Definitely not using `echo` to create it.

File: io.github.picocrypt.Picocrypt.yaml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      - install -Dm755 Picocrypt /app/bin/picocrypt-gui
+      - install -Dm644 key.svg /app/share/icons/hicolor/scalable/apps/io.github.picocrypt.Picocrypt.svg
+      - install -Dm644 io.github.picocrypt.Picocrypt.desktop /app/share/applications/io.github.picocrypt.Picocrypt.desktop
+      - install -Dm644 io.github.picocrypt.Picocrypt.metainfo.xml /app/share/metainfo/io.github.picocrypt.Picocrypt.metainfo.xml
+    sources:
+      - type: file
+        url: https://github.com/Picocrypt/Picocrypt/releases/download/1.42/Picocrypt
```
Comment (hfiguiere) [inline io.github.picocrypt.Picocrypt.yaml:30]: you should build it from source.

File: io.github.picocrypt.Picocrypt.metainfo.xml [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    </ul>
+  </description>
+
+  <categories>
+    <category>Utility</category>
+    <category>Security</category>
+  </categories>
+
+  <launchable type="desktop-id">io.github.picocrypt.Picocrypt.desktop</launchable>
+  <icon type="stock">io.github.picocrypt.Picocrypt</icon>
```
Comment (hfiguiere) [inline io.github.picocrypt.Picocrypt.metainfo.xml:40]:
```suggestion
```

File: io.github.picocrypt.Picocrypt.metainfo.xml
[the remark this answers is not in the document]
Comment (Retengart) [reply io.github.picocrypt.Picocrypt.metainfo.xml:1]: What is meant by "submitted upstream"?

File: flathub.json (hunk 1 above)
Comment (HACKERALERT) [reply flathub.json:2]: The binary is for x86_64, I also don't have an aarch64 machine to test if I did make aarch64 binaries.

File: io.github.picocrypt.Picocrypt.metainfo.xml
[the remark this answers is not in the document]
Comment (Retengart) [reply io.github.picocrypt.Picocrypt.metainfo.xml:1]: > What is meant by "submitted upstream"?

What is meant by "submitted upstream"?

[the remark this answers is not in the document]
Comment (HACKERALERT) [reply io.github.picocrypt.Picocrypt.metainfo.xml:1]: Upstream where?

File: io.github.picocrypt.Picocrypt.yaml (hunk 4 above)
Comment (Retengart) [reply io.github.picocrypt.Picocrypt.yaml:30]: > you should build it from source.

This is not very convenient, we will have to put the Golang inside for building. But the binary is already self-sufficient in itself, what prevents it from being used 'as is'?

File: io.github.picocrypt.Picocrypt.metainfo.xml
[the remark this answers is not in the document]
Comment (hfiguiere) [reply io.github.picocrypt.Picocrypt.metainfo.xml:1]: upstream as wherever the development is done. Where the original source is.

Here, it is **down**stream

File: flathub.json (hunk 1 above)
Comment (hfiguiere) [reply flathub.json:2]: hence the "build from source".

File: io.github.picocrypt.Picocrypt.yaml (hunk 4 above)
Comment (hfiguiere) [reply io.github.picocrypt.Picocrypt.yaml:30]: why is it not convenient? It's better.

File: io.github.picocrypt.Picocrypt.metainfo.xml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  <name>Picocrypt</name>
+  <summary>Small, simple, secure encryption tool</summary>
+  <developer id="io.github.picocrypt">
+    <name>Evan Su</name>
+  </developer>
+  <metadata_license>CC0-1.0</metadata_license>
+  <project_license>GPL-3.0</project_license>
+  <url type="homepage">https://github.com/Picocrypt/Picocrypt</url>
+  <url type="bugtracker">https://github.com/Picocrypt/Picocrypt/issues</url>
+  <content_rating type="oars-1.0" />
```
Comment (bbhtt) [inline io.github.picocrypt.Picocrypt.metainfo.xml:13]:
```suggestion
  <content_rating type="oars-1.1" />
```

File: io.github.picocrypt.Picocrypt.metainfo.xml [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  <categories>
+    <category>Utility</category>
+    <category>Security</category>
+  </categories>
+
+  <launchable type="desktop-id">io.github.picocrypt.Picocrypt.desktop</launchable>
+  <icon type="stock">io.github.picocrypt.Picocrypt</icon>
+  <provides>
+    <id>io.github.picocrypt.Picocrypt.desktop</id>
+  </provides>
```
Comment (bbhtt) [inline io.github.picocrypt.Picocrypt.metainfo.xml:43]:
```suggestion
```

File: io.github.picocrypt.Picocrypt.metainfo.xml [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  <screenshots>
+    <screenshot type="default">
+      <image>https://github.com/Picocrypt/Picocrypt/raw/main/images/screenshot.png</image>
```
Comment (bbhtt) [inline io.github.picocrypt.Picocrypt.metainfo.xml:46]:
```suggestion
      <image>https://github.com/Picocrypt/Picocrypt/raw/4d4bd47efe88ff25f372db81c4249920d399226b/images/screenshot.png</image>
```

File: io.github.picocrypt.Picocrypt.yaml (hunk 4 above)
Comment (bbhtt) [reply io.github.picocrypt.Picocrypt.yaml:30]: > This is not very convenient, we will have to put the Golang inside for building.

we have a go extension to take care of the toolchain https://github.com/flathub/org.freedesktop.Sdk.Extension.golang

See how other go apps do it https://github.com/flathub/dev.skynomads.Seabird/blob/master/dev.skynomads.Seabird.yaml

Comment (Retengart) [reply io.github.picocrypt.Picocrypt.yaml:30]: > > This is not very convenient, we will have to put the Golang inside for building.
> 
> we have a go extension to take care of the toolchain https://github.com/flathub/org.freedesktop.Sdk.Extension.golang
> 
> See how other go apps do it https://github.com/flathub/dev.skynomads.Seabird/blob/master/dev.skynomads.Seabird.yaml

But how do we handle dependencies if there is no network in flatpak environment during building? 
`require (
	github.com/Picocrypt/dialog v0.0.0-20240831001746-9ca708a9cd29
	github.com/Picocrypt/giu v0.0.0-20240831005244-5771b35043ac
	github.com/Picocrypt/imgui-go v0.0.0-20240831004007-6f60d7beadf6
	github.com/Picocrypt/infectious v0.0.0-20240830233326-3a050f65f9ec
	github.com/Picocrypt/serpent v0.0.0-20240830233833-9ad6ab254fd7
	github.com/Picocrypt/zxcvbn-go v0.0.0-20240831000415-fccb38ccb913
	golang.org/x/crypto v0.26.0
)

require (
	github.com/Picocrypt/gl v0.0.0-20240831002619-6531d2bba5fc // indirect
	github.com/Picocrypt/glfw/v3.3/glfw v0.0.0-20240831003212-7f16c5fb374b // indirect
	github.com/Picocrypt/mainthread v0.0.0-20240831004314-496f638392b3 // indirect
	github.com/Picocrypt/w32 v0.0.0-20240831001500-1183079d4d57 // indirect
	golang.org/x/sys v0.24.0 // indirect
)
`

**Comment by HACKERALERT** [comment]
Is building from source an absolute requirement? Because the existing binary is already done by Actions, I don't see why we can't just bundle it because building from source doesn't do anything different, it just repeats what the existing workflow to generate the binaries does.

**Comment by hfiguiere** [comment]
>  building from source doesn't do anything different

It would build on aarch64 too.

**Comment by HACKERALERT** [comment]
Fair. Is aarch64 a requirement? Because I've never tested it on aarch64 so I don't know if it even compiles ;)

File: io.github.picocrypt.Picocrypt.yaml (hunk 4 above)
Comment (bbhtt) [reply io.github.picocrypt.Picocrypt.yaml]: Use the go-vendor generator script to get a dependency manifest of your go modules then add that manifest to the PR

https://github.com/flatpak/flatpak-builder-tools/blob/master/go-get/flatpak-go-vendor-generator.py

**Comment by Retengart** [comment]
Finally managed to build from sources via Flatpak manifest, so there will be a new comit soon

File: io.github.picocrypt.Picocrypt.yaml (hunk 4 above)
Comment (Retengart) [reply io.github.picocrypt.Picocrypt.yaml]: > Use the go-vendor generator script to get a dependency manifest of your go modules then add that manifest to the PR
> 
> https://github.com/flatpak/flatpak-builder-tools/blob/master/go-get/flatpak-go-vendor-generator.py

Thanks, helped a lot

**Comment by HACKERALERT** [comment]
Synced the branch, this PR should be up to date now.

File: io.github.picocrypt.Picocrypt.yaml [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+  - --socket=x11
+  - --device=dri
+  - --filesystem=home
+build-options:
+  env:
+    - GOBIN=/app/bin
+    - GOROOT=/usr/lib/sdk/golang
+    - CGO_ENABLED=1
+  build-args:
+    - --share=network
```
Comment (bbhtt) [inline io.github.picocrypt.Picocrypt.yaml:19]: This is not allowed on Flathub. You have to add the generated manifest instead. This is for generating that manifest "locally"

Comment (Retengart) [reply io.github.picocrypt.Picocrypt.yaml:19]: > This is not allowed on Flathub. You have to add the generated manifest instead. This is for generating that manifest "locally"

We have taken this into account, there will be a new commit soon

**Comment by HACKERALERT** [comment]
Updated again

**Comment by bbhtt** [comment]
bot, build io.github.picocrypt.Picocrypt

**Comment by bbhtt** [comment]
Can you put desktop and metainfo file and use it from there ?

**Comment by HACKERALERT** [comment]
Put it where?

**Comment by bbhtt** [comment]
> Put it where?

https://github.com/Picocrypt/Picocrypt/

**Comment by HACKERALERT** [comment]
In a particular folder or at the root?

**Comment by bbhtt** [comment]
Doesn't matter, ideally it should be included in the tarball in github releases. But for now you can put it anywhere.

Then remove them from here and use it as:

```
- type: file
  url: https://raw.githubusercontent.com/...
  sha256: ...
```

**Comment by HACKERALERT** [comment]
Cool, done.

File: io.github.picocrypt.Picocrypt.yaml [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+      - install -Dm00644 $FLATPAK_ID.metainfo.xml $FLATPAK_DEST/share/appdata/$FLATPAK_ID.metainfo.xml
+    sources:
+      - type: archive
+        url: "https://github.com/Picocrypt/Picocrypt/archive/refs/tags/1.42.tar.gz"
+        sha256: b3c577d39a4f1a87131a2013894fb0b510cf147677b91d0f58313f0c85717474
+      - type: file
+        url: https://raw.githubusercontent.com/Picocrypt/Picocrypt/main/dist/flatpak/io.github.picocrypt.Picocrypt.metainfo.xml
+        sha256: a3a555bf36dc17be6549539c9837d963267bd58ba4805a0eb04cedaba11864c4
+      - type: file
+        url: https://raw.githubusercontent.com/Picocrypt/Picocrypt/main/dist/flatpak/io.github.picocrypt.Picocrypt.desktop
```
Comment (bbhtt) [inline io.github.picocrypt.Picocrypt.yaml:36]:
```suggestion
        url: https://raw.githubusercontent.com/Picocrypt/Picocrypt/e8016064e8136429e80e98fd402e82a0ada6821f/dist/flatpak/io.github.picocrypt.Picocrypt.metainfo.xml
        sha256: a3a555bf36dc17be6549539c9837d963267bd58ba4805a0eb04cedaba11864c4
      - type: file
        url: https://raw.githubusercontent.com/Picocrypt/Picocrypt/e8016064e8136429e80e98fd402e82a0ada6821f/dist/flatpak/io.github.picocrypt.Picocrypt.desktop
```

**Comment by HACKERALERT** [comment]
Applied

**Comment by bbhtt** [comment]
bot, build io.github.picocrypt.Picocrypt

**Comment by HACKERALERT** [comment]
Also is it a good idea to bump sdk to 24.08?

**Comment by bbhtt** [comment]
Yea sure

**Comment by HACKERALERT** [comment]
Dropped back down to 23.08 because build failed, likely 24.08 is too new currently.

**Comment by HACKERALERT** [comment]
bot, build io.github.picocrypt.Picocrypt

**Comment by bbhtt** [comment]
If you are using `home` for drag and drop, you might want to look into the Filetransfer portal https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.FileTransfer.html which eliminates the need for that and enables complete filesystem sandboxing.

**Comment by HACKERALERT** [comment]
It also has custom open and save dialogs, which is not drag and drop and requires home access probably. The filetransfer portal also seems ad-hoc and would require platform-specific code which is not ideal.

**Comment by Retengart** [comment]
I've tested the build, everything works fine

**Pull request closed without merging**